include/multipass/cli/format_utils.h

```diff
@@ -33,44 +34,77 @@ class Formatter;
 // which takes as input the element in the range and returns its width in columns.
-static constexpr auto column_width = [](const auto begin, const auto end, const auto get_width, int minimum_width,
-                                        int space = 2) {
-    if (0 == std::distance(begin, end))
-        return minimum_width;
-
-    auto max_width =
-        std::max_element(begin, end, [&get_width](auto& lhs, auto& rhs) { return get_width(lhs) < get_width(rhs); });
-    return std::max(get_width(*max_width) + space, minimum_width);
-};
+static constexpr auto column_width =
+    [](const auto begin, const auto end, const auto get_width, int header_width, int minimum_width = 0) {
+        if (0 == std::distance(begin, end))
+            return std::max({header_width + col_buffer, minimum_width});
+
+        auto max_width = std::max_element(begin, end, [&get_width](auto& lhs, auto& rhs) {
+            return get_width(lhs) < get_width(rhs);
+        });
+        return std::max({get_width(*max_width) + col_buffer, header_width + col_buffer, minimum_width});
+    };
 } // namespace format
 } // namespace multipass
 
-template <typename Instances>
-Instances multipass::format::sorted(const Instances& instances)
+template <typename Container>
+Container multipass::format::sorted(const Container& items)
 {
-    if (instances.empty())
-        return instances;
+    if (items.empty())
+        return items;
 
-    auto ret = instances;
+    auto ret = items;
     const auto petenv_name = MP_SETTINGS.get(petenv_key).toStdString();
     std::sort(std::begin(ret), std::end(ret), [&petenv_name](const auto& a, const auto& b) {
-        if (a.name() == petenv_name)
+        using T = std::decay_t<decltype(a)>;
+        using google::protobuf::util::TimeUtil;
+
+        // Put instances first when sorting info reply
+        if constexpr (std::is_same_v<T, multipass::DetailedInfoItem>)
+        {
+            if (a.has_instance_info() && b.has_snapshot_info())
+                return true;
+            else if (a.has_snapshot_info() && b.has_instance_info())
+                return false;
+        }
+
+        // Put petenv related entries first
+        if (a.name() == petenv_name && b.name() != petenv_name)
             return true;
-        else if (b.name() == petenv_name)
+        else if (b.name() == petenv_name && a.name() != petenv_name)
             return false;
         else
+        {
+            // Sort by timestamp when names are the same for snapshots
+            if constexpr (std::is_same_v<T, multipass::DetailedInfoItem>)
+            {
+                if (a.has_snapshot_info() && a.name() == b.name())
+                    return TimeUtil::TimestampToNanoseconds(a.snapshot_info().fundamentals().creation_timestamp()) <
+                           TimeUtil::TimestampToNanoseconds(b.snapshot_info().fundamentals().creation_timestamp());
```

Comment:
I didn't realize it was taken in milliseconds. It might be better to take them in nanoseconds, to better protect against parallel requests. It would be a corner case, with only mild detriment (bad sorting of output). But I agree we should align the resolution we use on both sides, so if it is just an easy type change on the daemon, we could attempt that. If we find it would take longer for any reason, I would advise leaving it for now and replacing with a TODO.

Reply:
This is fine after the explanation.